### PR TITLE
feat: embedded web UI — xv ui

### PR DESCRIPTION
**Pull request by bziobnic**
## Summary

Adds an embedded, localhost-only web UI behind a new `ui` cargo feature. `xv ui` starts an axum server bound to 127.0.0.1 (ephemeral port by default), opens the browser, and serves a zero-build vanilla-JS single-page app plus a JSON API that delegates to the existing backend trait layer — so all backends (Azure, AWS, local/offline) work with no logic reimplemented.

- **Secrets:** list/search, create, reveal/copy, edit (value + folder/groups/note/expiry, preserving content-type and custom tags), rename/move, delete
- **Files:** drag-and-drop upload (100 MB cap), download (RFC 5987 filenames), delete — hidden when the backend has no file storage
- **Vaults:** switcher in the header; capability-aware UI via `/api/context`
- **Security:** loopback bind only, per-session 256-bit bearer token (in-memory on the page, scrubbed from the URL), Host/Origin validation with strict port parsing, `Cache-Control: no-store` on every API response including auth rejections, secret values never in URLs
- **Footprint:** one new dependency (`axum`, optional, feature-gated); non-`ui` builds are unchanged

Design spec: `docs/superpowers/specs/2026-07-08-web-ui-design.md` · User docs: `docs/web-ui.md`

## Test plan

- 18 `src/web` tests: auth (token, Host/Origin, no-store on rejections), secret CRUD/reveal/move roundtrips, file multipart upload/download/delete, >2 MB upload, non-ASCII and quoted filenames, metadata preservation on edit
- Full suite: 687 passing with `--features ui`, 673 without; clippy clean in `src/web`
- Live E2E smoke against the offline local backend: token-less request → 401 + no-store, spoofed Host → 403, list/reveal/create correct, assets served

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> New loopback HTTP surface exposes secret CRUD and file I/O; mitigations are solid but any auth bypass or Host parsing bug could expose vault data to malicious pages on the same machine.
> 
> **Overview**
> Introduces **`xv ui`** behind a new **`ui` cargo feature** (optional **axum**), wiring a **`src/web/`** module into the CLI like the existing TUI gate.
> 
> The server binds **127.0.0.1** only, issues a **per-session bearer token** (URL query on first open, then in-memory in the page), and protects **`/api/*`** with **Host/Origin checks**, **`Cache-Control: no-store`**, and handlers that delegate to **`Arc<dyn Backend>`** — no duplicated vault logic. JSON endpoints cover **context/vaults**, **secret list/CRUD/reveal/move**, and **file list/upload/download/delete** when `file-ops` is enabled (multipart uploads with a raised body limit).
> 
> The **embedded** `index.html` / `app.js` / `style.css` provide secrets table + edit drawer (reveal/copy, metadata edits that preserve **content-type**, custom tags, **enabled**, **not_before**) and a files tab with drag-and-drop. **`docs/web-ui.md`**, README/CLAUDE/CHANGELOG updates, plus a large implementation plan/spec under `docs/superpowers/`.
> 
> **Tests:** extensive `src/web` integration tests (auth, secret roundtrips, file edge cases). Default builds without `--features ui` stay axum-free.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 89d0f4998ec71bfa51dbaeee204b7c0a3d8b8960. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/web/api.rs [hunk 1]
```diff
@@ -0,0 +1,856 @@
...
+        value: Zeroizing::new(body.value),
+        content_type: body.content_type,
+        enabled: Some(true),
+        expires_on: body.expires_on,
+        not_before: body.not_before,
+        tags: body.tags,
+        groups: body.groups,
+        note: body.note,
+        folder: body.folder,
+    };
```
[the remark this answers is not in the document]
Comment (bziobnic) [reply src/web/api.rs:184]: Fixed in 12a2c33 — `PutSecretBody` now carries `enabled` (absent defaults to enabled for new secrets) and the drawer echoes the secret's current enabled state on value edits, so editing a disabled secret no longer re-enables it. Regression test: `put_preserves_enabled_and_not_before_across_edits`.

[the remark this answers is not in the document]
Comment (bziobnic) [reply src/web/api.rs:184]: Fixed in 12a2c33 — the drawer now stashes `not_before` from the metadata fetch and echoes it in the PUT body (alongside content_type/custom tags/enabled), so value edits preserve the not-before constraint. Covered by the same regression test.

File: src/web/assets/app.js [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+  const { vaults } = await api('GET', '/api/vaults');
+  const sel = $('#vault-select');
+  sel.innerHTML = '';
+  for (const v of vaults) {
+    const opt = document.createElement('option');
+    opt.value = opt.textContent = v.name;
+    opt.selected = v.name === currentVault;
+    sel.appendChild(opt);
+  }
+  sel.onchange = () => { currentVault = sel.value; loadSecrets(); loadFiles(); };
```
[the remark this answers is not in the document]
Comment (bziobnic) [reply src/web/assets/app.js]: Fixed in cbdd600 — switching vaults now closes the drawer (new `closeDrawer()` resets `editing`/`editingMeta`), so drawer actions can no longer target the same name in the newly selected vault.

File: src/web/assets/app.js [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        value: f.value.value,
+        folder: f.folder.value || null,
+        note: f.note.value || null,
+        groups: groups.length ? groups : null,
+        expires_on: f.expires_on.value || null,
+        content_type: editingMeta?.content_type || null,
+        tags: editingMeta && Object.keys(editingMeta.tags).length ? editingMeta.tags : null,
+        enabled: editingMeta ? editingMeta.enabled : true,
+        not_before: editingMeta?.not_before || null,
+      });
```
[the remark this answers is not in the document]
Comment (bziobnic) [reply src/web/assets/app.js:189]: Fixed in cbdd600 — if the metadata fetch fails, the drawer no longer opens (error toast, `editing` reset), so a save can't send `enabled: true` / empty tags for a secret whose real state was never loaded.

File: src/web/mod.rs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    let mut buf = [0u8; 32];
+    rand::rng().fill_bytes(&mut buf);
+    let token = hex::encode(buf);
+
+    let state = Arc::new(WebState {
+        backend,
+        token: token.clone(),
+        vault,
+    });
```
[the remark this answers is not in the document]
Comment (bziobnic) [reply src/web/mod.rs:112]: Scoped as a documented limitation rather than fixed here: `resolve_vault_for_trait` intentionally survived the workspace refactor as the plain default-vault resolver (see its comment — same path `xv gen`/`find --all-vaults` use), and the dropdown only lists the active backend's real vaults, which round-trip to that same backend — so there's no cross-backend mutation path from the UI's own vault list. Multi-backend workspace/alias resolution (the `resolve_workspace_or_default` seam, as the TUI uses) is real missing functionality though — documented in docs/web-ui.md and tracked as a follow-up issue.
